File: receipt-analyzer-service/src/main/java/pl/madzierski/daniel/app/product_dict/product_category/ProductCategoryService.java
package pl.madzierski.daniel.app.product_dict.product_category;

import lombok.AllArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import pl.madzierski.daniel.app.product_dict.ProductDictRepository;
import pl.madzierski.daniel.app.product_dict.product_category.model.*;
import pl.madzierski.daniel.exception.AppRuntimeException;
import pl.madzierski.daniel.exception.AppRuntimeExceptionMessages;

import java.util.List;

@Service
@AllArgsConstructor
class ProductCategoryService {

    private final ProductCategoryRepository productCategoryRepository;
    private final ProductDictRepository productDictRepository;

    public GetProductCategoryListResponse getProductCategoryList() {
        List<GetProductCategoryListResponse.ProductCategory> items = productCategoryRepository.findAllByOrderByNameAsc().stream()
                .map(productCategory -> new GetProductCategoryListResponse.ProductCategory(productCategory.getId(), productCategory.getName()))
                .toList();
        return new GetProductCategoryListResponse(items);
    }

    @Transactional
    public CreateProductCategoryResponse addProductCategory(CreateProductCategoryRequest request) {
        String name = request.name().trim();
        if (productCategoryRepository.existsByName(name)) {
            throw new AppRuntimeException(AppRuntimeExceptionMessages.PRODUCT_CATEGORY_ALREADY_EXISTS);
        }

        ProductCategoryEntity savedProductCategory = productCategoryRepository.save(new ProductCategoryEntity(name));
        return new CreateProductCategoryResponse(savedProductCategory.getId(), savedProductCategory.getName());
    }

    @Transactional
    public UpdateProductCategoryResponse updateProductCategory(String productCategoryId, UpdateProductCategoryRequest request) {
        ProductCategoryEntity productCategory = productCategoryRepository.findById(productCategoryId)
                .orElseThrow(() -> new AppRuntimeException(AppRuntimeExceptionMessages.PRODUCT_CATEGORY_NOT_FOUND));
        String name = request.name().trim();

        if (!productCategory.getName().equals(name) && productCategoryRepository.existsByNameAndIdNot(name, productCategoryId)) {
            throw new AppRuntimeException(AppRuntimeExceptionMessages.PRODUCT_CATEGORY_ALREADY_EXISTS);
        }

        productCategory.setName(name);
        ProductCategoryEntity savedProductCategory = productCategoryRepository.save(productCategory);
        return new UpdateProductCategoryResponse(savedProductCategory.getId(), savedProductCategory.getName());
    }

    @Transactional
    public void deleteProductCategory(String productCategoryId) {
        ProductCategoryEntity productCategory = productCategoryRepository.findById(productCategoryId)
                .orElseThrow(() -> new AppRuntimeException(AppRuntimeExceptionMessages.PRODUCT_CATEGORY_NOT_FOUND));
        if (productDictRepository.countByProductCategory_Id(productCategory.getId()) > 0) {
            throw new AppRuntimeException(AppRuntimeExceptionMessages.PRODUCT_CATEGORY_IN_USE);
        }
        productCategoryRepository.delete(productCategory);
    }

}

File: web-client/src/app/service/product-category.service.ts
import {inject, Injectable} from '@angular/core';
import {HttpClient} from '@angular/common/http';
import {Observable} from 'rxjs';
import {
  CreateProductCategoryRequest,
  CreateProductCategoryResponse,
  GetProductCategoryListResponse,
  UpdateProductCategoryRequest,
  UpdateProductCategoryResponse
} from '../model/product-category.model';

@Injectable({
  providedIn: 'root',
})
export class ProductCategoryService {
  private readonly httpClient = inject(HttpClient);

  getProductCategoryList(): Observable<GetProductCategoryListResponse> {
    return this.httpClient.get<GetProductCategoryListResponse>(`/api/product-categories`);
  }

  createProductCategory(request: CreateProductCategoryRequest): Observable<CreateProductCategoryResponse> {
    return this.httpClient.post<CreateProductCategoryResponse>(`/api/product-categories`, request);
  }

  updateProductCategory(id: string, request: UpdateProductCategoryRequest): Observable<UpdateProductCategoryResponse> {
    return this.httpClient.put<UpdateProductCategoryResponse>(`/api/product-categories/${id}`, request);
  }

  deleteProductCategory(id: string): Observable<void> {
    return this.httpClient.delete<void>(`/api/product-categories/${id}`);
  }
}
